Launch Elasticsearch and Kibana automatically

satnam6502 opened the pull request:
This PR changes our setup to automatically launch an Elasticsearch pod for ingesting logs and a Kibana pod for viewing logs when logging with Fluentd to Elasticsearch is selected. The URLs for Elasticsearch and the Kibana dashboard are reported at cluster creation time:

```
.. setting up Fluentd to Elasticsearch logging with Kibana viewer
Created [https://www.googleapis.com/compute/v1/projects/kubernetes-elk/global/firewalls/fluentd-elasticsearch-logging].
NAME                          NETWORK SRC_RANGES RULES                      SRC_TAGS TARGET_TAGS
fluentd-elasticsearch-logging default 0.0.0.0/0  tcp:5601,tcp:9200,tcp:9300          kubernetes-minion

Cluster logs are ingested into Elasticsearch running at http://130.211.126.68:9200
Kibana logging dashboard will be available at http://23.236.59.213:5601
```

I have also changed the names of the Elasticsearch and Kibana pods and services (adding -logging) to avoid any name clashes with other instances of Elasticsearch in the cluster.

satnam6502 committed:
Launch Elasticsearch and Kibana automatically

thockin commented on cluster/addons/fluentd-elasticsearch/es-pod.yml:
```diff
@@ -0,0 +1,25 @@
+apiVersion: v1beta1
+kind: Pod
```
These need to run under replication controllers, not bare pods.  Monitoring is wrong here, too @vishh #2408

satnam6502 replied:
Yes, I had intended to change this to have a replication controller (still with one controlled pod). I'll make the change.

vishh replied:
@thockin: I am working on replication controllers. I will add them along with a few other enhancements to monitoring.

thockin commented on cluster/gce/util.sh:
```diff
@@ -783,6 +783,35 @@ function teardown-monitoring {
+function setup-fluentd-elasticsearch-logging {
+  echo .. setting up Fluentd to Elasticsearch logging with Kibana viewer
+  local -r kubectl="${KUBE_ROOT}/cluster/kubectl.sh"
+  if "${kubectl}" create -f "${KUBE_ROOT}"/cluster/addons/fluentd-elasticsearch/es-pod.yml &> /dev/null && \
```
We should decide on .yaml or .yml or .json and do the same thing everywhere.  DNS says .yaml, monitoring says .json, and you say .yml :)

satnam6502 replied:
When I wrote my first YAML file for this project I tried to fit in with what others had written and used ".yml" even though my fingers were programmed to type ".yaml". I am happy to change the suffixes if someone declares the standard. For human written files designed to be read by humans YAML seems like a much better choice than JSON.

thockin replied:
I agree - @vishh should change his json to yaml.  I like .yaml, fwiw, and
it is more prevalent in our tree

On Wed, Jan 7, 2015 at 4:20 PM, Satnam Singh <notifications@github.com>
wrote:

> In cluster/gce/util.sh
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3292#discussion-diff-22627653>
> :
>
> > @@ -783,6 +783,35 @@ function teardown-monitoring {
> >    fi
> >  }
> >
> > +function setup-fluentd-elasticsearch-logging {
> > +  echo .. setting up Fluentd to Elasticsearch logging with Kibana viewer
> > +  local -r kubectl="${KUBE_ROOT}/cluster/kubectl.sh"
> > +  if "${kubectl}" create -f "${KUBE_ROOT}"/cluster/addons/fluentd-elasticsearch/es-pod.yml &> /dev/null && \
>
> When I wrote my first YAML file for this project I tried to fit in with
> what others had written and used ".yml" even though my fingers were
> programmed to type ".yaml". I am happy to change the suffixes if someone
> declares the standard. For human written files designed to be read by
> humans YAML seems like a much better choice than JSON.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3292/files#r22627653>
> .
>

vishh replied:
Ack. Will change the format in a separate PR.

satnam6502 commented:
@thockin : changed to using replication controllers, all yml files changed to yaml and updated documentation. PTAL.

satnam6502 commented:
I just did one more tweak to add tag version numbers to the two Dockerfiles that we create (both at tag 1.0 here) and then I mention this tag in the pod specs. This will allow us to update the logging work without breaking existing clusters.

thockin commented on cluster/aws/config-default.sh:
```diff
@@ -38,6 +38,7 @@ PORTAL_NET="10.0.0.0/16"
 
 # Optional: Install node monitoring.
 ENABLE_NODE_MONITORING=true
+ELASTICSEARCH_LOGGING_REPLICAS=1
```
Should probably be lumped with other logging flags, rather than with node-monitoring

satnam6502 replied:
Oops, my mistake, fixed.

thockin commented on cluster/gce/config-default.sh:
```diff
@@ -53,6 +53,7 @@ ENABLE_DOCKER_REGISTRY_CACHE=true
 # Optional: Enable node logging.
 ENABLE_NODE_LOGGING=true
 LOGGING_DESTINATION=elasticsearch # options: elasticsearch, gcp
+ELASTICSEARCH_LOGGING_REPLICAS=1
```
You should guard this with ENABLE_CLUSTER_LOGGING, similar to how monitoring is split.  I know it's a plumbing mess...

satnam6502 replied:
Done.

thockin commented on cluster/kube-up.sh:
```diff
@@ -54,4 +54,13 @@ if [[ "${ENABLE_CLUSTER_DNS}" == "true" ]]; then
+# If logging with Fluentd to Elasticsearch is enabled then create pods
+# and services for Elasticsearch (for ingesting logs) and Kibana (for
+# viewing logs).
+if [[ "${ENABLE_NODE_LOGGING-}" == "true" ]]; then
```
I think it's worthwhile to split ENABLE and setup for on-node logging infra from in-cluster logging infra.

thockin commented:
If we use this at cluster setup, it needs to move to cluster/addons

zmerlynn commented:
This PR is going to cause a headache for GKE. Please make sure to be delicate around the GKE provider on the OSS side or you'll break GKE Jenkins. Check it as disable initially there, create a bug, assign it to me. I'll consider how to generically launch other pods when I'm looking at #3305.

satnam6502 commented:
For GKE, can we just set `ENABLE_NODE_LOGGING=false` ? Is that not good enough?

zmerlynn commented:
Yeah, you can just set it to false for now. It looks like it's default off if you don't plumb the variable through? (Or did I misread the PR?)

You can just "export KUBERNETES_PROVIDER=gke" and then "go run hack/e2e.go -up" etc. to get gke.

satnam6502 commented:
@thockin : I think I have done your bidding. PTAL.
@zmerlynn : I will check the GKE provider to make sure I have not broken anything.
All other provider maintainers please note changes to `config-default.sh`
I also added a few missing `--project` flags which are needed when the project is set in `config-default.sh` @cjcullen @jlowdermilk @mbforbes

satnam6502 commented:
@zmerlynn : running the e2e tests with the GKE provider did not give an encouraging result. Running with just `-up` works.
```
$ hack/e2e-test.sh
....
2015/01/08 19:55:48 Passed tests: certs.sh[1/1] goe2e.sh[1/1] monitoring.sh[1/1]
2015/01/08 19:55:48 Flaky tests:
2015/01/08 19:55:48 Failed tests: basic.sh[0/1] guestbook.sh[0/1] liveness.sh[0/1] pd.sh[0/1] private.sh[0/1] services.sh[0/1] update.sh[0/1]
2015/01/08 19:55:48 7 test(s) failed.
2015/01/08 19:55:48 Running: teardown
...
```

thockin commented on cluster/aws/config-default.sh:
```diff
@@ -36,12 +36,20 @@ MINION_SCOPES=""
+ENABLE_CLUSTER_LOGGING=false
+if [[ "${ENABLE_NODE_LOGGING-}" == "true" ]]; then
+  if [[ "${LOGGING_DESTINATION-}" == "elasticsearch" ]]; then
+    if [[ "${ENABLE_CLUSTER_LOGGING-}" == "true" ]]; then
```
Do you really need all the logic, or can you just set the variable, or even just set it and comment the line out?

satnam6502 replied:
I think I added that logic because you wanted the `ELASTICSEARCH_LOGGING_REPLICAS` definition guarded? Initially, I did just set the variable.

thockin replied:
Hmm, I don't recall saying that.  Maybe something I said was too cryptic?  Either way, I think this should be a config file, not a program.

satnam6502 replied:
I am sure I just mis-understood you. Shall I remove two of the three tests, or just set the variable at the top-level like I did originally?

thockin replied:
I think simpler is better in the config files.

On Thu, Jan 8, 2015 at 8:40 PM, Satnam Singh <notifications@github.com>
wrote:

> In cluster/aws/config-default.sh
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3292#discussion-diff-22701289>
> :
>
> >  LOGGING_DESTINATION=elasticsearch # options: elasticsearch, gcp
> > +
> > +# Optional: When set to true, Elasticsearch and Kibana will be setup as part of the cluster bring up.
> > +ENABLE_CLUSTER_LOGGING=false
> > +if [[ "${ENABLE_NODE_LOGGING-}" == "true" ]]; then
> > +  if [[ "${LOGGING_DESTINATION-}" == "elasticsearch" ]]; then
> > +    if [[ "${ENABLE_CLUSTER_LOGGING-}" == "true" ]]; then
>
> I am sure I just mis-understood you. Shall I remove two of the three
> tests, or just set the variable at the top-level like I did originally?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3292/files#r22701289>
> .
>

thockin commented on cluster/kube-down.sh:
```diff
@@ -28,6 +28,18 @@ echo "Bringing down cluster using provider: $KUBERNETES_PROVIDER"
+# If logging with Fluentd to Elasticsearch was enabled then
+# delete the pods and service and remove the firewall rule
+# for accessing Elasticsearch and Kibana.
+if [[ "${ENABLE_NODE_LOGGING-}" == "true" ]]; then
```
Simpler to just call teardown-logging and embed the logic in that function?

satnam6502 replied:
I thought it made sense to have the predicate outside since the function that "tears down logging" because I thought this should only be called if logging was enabled in the first place. OK: I shall push the predicate into the function (which makes it similar in style to how the monitoring code is written). But I am sure some software engineering pedant would have something to say about this...

thockin replied:
I prefer to think of the polymorphic nature of it - a platform might define a different implementation of setup-logging - that is the API.  fluent/elastic is the impl

satnam6502 replied:
Agreed, and as I said, I'll make the change. Although I am not sure if that's an approved use of the term *polymorphic* :-) Perhaps that's because I assume "parametric" in front of the word but others assume "ad hoc".

satnam6502 commented:
@thockin : moved predicate to inside the bring up and tear down functions for cluster logging; removed conditional guards for setting `ELASTICSEARCH_LOGGING_REPLICAS`. Tested. PTAL.

zmerlynn commented:
@satnam: Sorry, to test on GKE you'll also need --check_version_skew=false.

On Thu, Jan 8, 2015 at 8:01 PM, Satnam Singh <notifications@github.com>
wrote:

> @zmerlynn <https://github.com/zmerlynn> : running the e2e tests with the
> GKE provider did not give an encouraging result. Running with just -up
> works.
>
> $ hack/e2e-test.sh
> ....
> 2015/01/08 19:55:48 Passed tests: certs.sh[1/1] goe2e.sh[1/1] monitoring.sh[1/1]
> 2015/01/08 19:55:48 Flaky tests:
> 2015/01/08 19:55:48 Failed tests: basic.sh[0/1] guestbook.sh[0/1] liveness.sh[0/1] pd.sh[0/1] private.sh[0/1] services.sh[0/1] update.sh[0/1]
> 2015/01/08 19:55:48 7 test(s) failed.
> 2015/01/08 19:55:48 Running: teardown
> ...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3292#issuecomment-69290437>
> .
>

satnam6502 commented:
Yes, I finally worked that out after trying various configurations and good to know about `--check_version_skew=false` for reducing the output verbiage. Thanks.

thockin commented on cluster/kube-up.sh:
```diff
@@ -54,4 +54,6 @@ if [[ "${ENABLE_CLUSTER_DNS}" == "true" ]]; then
       | "${KUBE_ROOT}/cluster/kubectl.sh" create -f -
 fi
 
+setup-fluentd-elasticsearch-logging
```
last nit: rename to setup-logging

satnam6502 commented:
Changed the bring up and tear down names to `setup-logging` and `teardown-loging`.
Merging with the approval of @thockin
